src/ssl/gadgets.cc

```diff
@@ -479,18 +479,24 @@ static bool buildCertificate(Security::CertPointer & cert, Ssl::CertificatePrope
         (void)replaceCommonName(cert, properties.commonName);
     }
 
-    // We should get caCert notBefore and notAfter fields and do not allow
-    // notBefore/notAfter values from certToMimic before/after notBefore/notAfter
-    // fields from caCert.
-    // Currently there is not any way in openssl tollkit to compare two ASN1_TIME
-    // objects.
+    ASN1_TIME *notBefore = NULL;
+    ASN1_TIME *notAfter = NULL;
+
+    // Get times from cacert
+    notBefore = X509_getm_notBefore(properties.signWithX509.get());
+    notAfter = X509_getm_notAfter(properties.signWithX509.get());
+
     ASN1_TIME *aTime = NULL;
     if (!properties.setValidBefore && properties.mimicCert.get())
         aTime = X509_getm_notBefore(properties.mimicCert.get());
     if (!aTime && properties.signWithX509.get())
```

Comment:
You want to check the presence of the limit in the CA certificate, not the presence of the CA certificate itself.

```suggestion
     if (!aTime && caNotBefore)
```

... but since assigning nil is OK here, perhaps we can simplify further:

```suggestion
     if (!aTime)
         aTime = caNotBefore; // may also be nil
```